hypervisor/arch/x86/guest/virtual_cr.c

```diff
@@ -34,23 +34,52 @@
 				CR4_OSXMMEXCPT | CR4_SMAP | CR4_PKE |        \
 				CR4_SMXE | CR4_UMIP)
 
+/* PAE PDPTE bits 1 ~ 2, 5 ~ 8 are always reserved */
+#define PAE_PDPTE_FIXED_RESVD_BITS	0x00000000000001E6
+
 static uint64_t cr0_always_on_mask;
 static uint64_t cr0_always_off_mask;
 static uint64_t cr4_always_on_mask;
 static uint64_t cr4_always_off_mask;
 
-static void load_pdptrs(const struct acrn_vcpu *vcpu)
+static int load_pdptrs(const struct acrn_vcpu *vcpu)
 {
 	uint64_t guest_cr3 = exec_vmread(VMX_GUEST_CR3);
-	/* TODO: check whether guest cr3 is valid */
-	uint64_t *guest_cr3_hva = (uint64_t *)gpa2hva(vcpu->vm, get_pae_pdpt_addr(guest_cr3));
-
-	stac();
-	exec_vmwrite64(VMX_GUEST_PDPTE0_FULL, get_pgentry(guest_cr3_hva + 0UL));
-	exec_vmwrite64(VMX_GUEST_PDPTE1_FULL, get_pgentry(guest_cr3_hva + 1UL));
-	exec_vmwrite64(VMX_GUEST_PDPTE2_FULL, get_pgentry(guest_cr3_hva + 2UL));
-	exec_vmwrite64(VMX_GUEST_PDPTE3_FULL, get_pgentry(guest_cr3_hva + 3UL));
-	clac();
+	struct cpuinfo_x86 *cpu_info = get_pcpu_info();
+	int ret = 0;
+	uint64_t pdpte[4]; /* Total four PDPTE */
+	uint64_t rsvd_bits_mask;
+	int maxphyaddr, i;
+
+	/* check whether the address area pointed by the guest cr3
+	 * can be accessed or not
+	 */
+	if (copy_from_gpa(vcpu->vm, pdpte, get_pae_pdpt_addr(guest_cr3), sizeof(pdpte)) != 0) {
+		ret = -EFAULT;
+	} else {
+		/* Check if any of the PDPTEs sets both the P flag
+		 * and any reserved bit
+		 */
+		maxphyaddr = cpu_info->phys_bits;
+		/* reserved bits: 1~2, 5~8, maxphyaddr ~ 63 */
+		rsvd_bits_mask = (63 < maxphyaddr) ? 0 : (((1ULL << (63 - maxphyaddr + 1)) - 1) << maxphyaddr);
```

Comment:
rsvd_bits_mask = (63 < maxphyaddr) ? 0UL : (((1UL << (63U - maxphyaddr + 1U)) - 1UL) << maxphyaddr);
maxphyaddr type shall be u32_t.